README.md
# [![Grav adsense Plugin](assets/logo.png)][project]

## The **Adsense** plugin for [Grav](https://getgrav.org) introduces integration of Adsense blocs for your modular page and your page content.

#### Table of Contents:

-   [About](#about)
-   [Installation and Updates](#installation-and-updates)
-   [configuration](#configuration)
-   [Usage](#usage)
-   [Contributing](#contributing)
-   [Credits](#credits)
-   [License](#license)

## About

`AdSense` is a plugin for [**Grav**](http://getgrav.org) that let you add Adsense blocs to some place in your modular page and/or page content. 

![Screenshot adsense Plugin](assets/screenshot.png "AdSense Preview")

For modular page, plugin allow user to define default `horizontal`, `vertical` or `square` Ad.

For page content, plugin allow user to define a list of Adsense blocs and integrate it through [Shortcode](https://github.com/getgrav/grav-plugin-shortcode-core).

The following kinds of Adsense are compatible:
- [Graphic and Text Ads](https://support.google.com/adsense/answer/185666?hl=en)
- [InArticle Ads](https://adsense.googleblog.com/2017/07/introducing-adsense-native-ads.html)
- [Auto Ads](https://adsense.googleblog.com/2018/02/introducing-adsense-auto-ads.html)

## Installation

Installing the `Adsense` plugin can be done in one of two way: GPM (Grav Package Manager) installation method enables you to quickly and easily install the plugin with a simple terminal command, while the manual method enables you to do so via a zip file or GIT.

### GPM Installation (Preferred)

The simplest way to install this plugin is via the [Grav Package Manager (GPM](http://learn.getgrav.org/advanced/grav-gpm) through your system's Terminal (also called the command line). From the root of your Grav install type:

    $ bin/gpm install adsense

This will install the `Adsense` plugin into your /user/plugins directory within Grav. Its files can be found under `user/plugins/adsense`.

### Manual Installation (Download)

To install this plugin, just download the zip version of this repository and unzip it under:

    user/plugins/adsense

## Configuration

Configuration can be done by the **Admin Panel** or by the configuration file `adsense.yaml` into your `users/config/plugins/` folder:

```yaml
# Global plugin configurations

enabled: true             # Set to false to disable this plugin completely
sandbox: false            # Enables a demo mode for local purpose
built_in_css: true        # Use the default plugin CSS
add_editor_button: true   # Add an Adsense button to Admin panel in order to add Adsense tag template
use_auto_ads: false       # Use Auto Ads in page (new feature of Google Adsense)

# Default configurations for AdSense

adsense:
  options:
    mode: "async"        # Code integration mode (either "async" or "sync")
    position: "center"   # Ad position in page (either "center", "left" or "right")
    auto_ads_client: ""     # Auto Ads client ID (like ca-pub-0000000000000000)
    auto_ads_filter:        # Filter of URN pages used to load Auto Ads. Nothing means All pages (Non filtered)
      - blog
      - portfolio
  page_ads:              # Ad blocs for page content
    -                    # List of ad blocs
      id: unique-id                         # Unique Id for the Ad
      client: ""                            # AdSense client (like ca-pub-0000000000000000)
      slot: 0000000000                      # AdSense slot
      type: inarticle                       # Adsense type (either "inarticle" or "normal")
      width: 468                            # Adsense width
      height: 60                            # Adsense height

  modular_ads:          # Ad blocs for modular page (as default)
    horizontal:         # For Horizontal Ad
      client: ""                            # AdSense client (like ca-pub-0000000000000000)
      slot: 0000000000                      # AdSense slot
      width: 728                            # Adsense width
      height: 90                            # Adsense height
    vertical:         # For Vertical Ad
      client: ""                            # AdSense client (like ca-pub-0000000000000000)
      slot: 0000000000                      # AdSense slot
      width: 200                            # Adsense width
      height: 600                           # Adsense height
    square:         # For Square Ad
      client: ""                            # AdSense client (like ca-pub-0000000000000000)
      slot: 0000000000                      # AdSense slot
      width: 300                            # Adsense width
      height: 300                           # Adsense height
```

## Usage

There are many ways to use `Adsense` plugin.

### Auto Ads

_Auto Ads_ is a new feature of *Google Adsense*. More informations are available here:

[https://adsense.googleblog.com/2018/02/introducing-adsense-auto-ads.html](https://adsense.googleblog.com/2018/02/introducing-adsense-auto-ads.html)

To use _Auto Ads_ in `Adsense` plugin, you just configure `use_auto_ads` to *true* and `adsense.options.auto_ads_client` to your *Auto Ads client ID*.

To add filtered URN items for Auto Ads treatment, please add items in parameter `adsense.options.auto_ads_filter`.  
When nothing is set, all pages load Auto Ads.  
When an item is set, URN with this item load Auto Ads.  
When a filter is set and you want to load Auto Ads for URN `/` ressource, please add the key `home`to filter list.

### By Shortcode

Using Shortcode allow user to integrate Adsense bloc in page content. There are 2 ways:

- Using unique ID

```md
[adsense id="my-id"][/adsense]
```

It look for the adsense bloc with unique ID  `id` in the configuration list `page_ads` and display the Ad.

- Using specific Adsense informations

```md
[adsense client="adsense-client" slot="adsense-slot" width="adsense-width" height="adsense-height" type="adsense-type"][/adsense]
```

It display the Ad with the tag informations.  
`type` can be :
- `inarticle` for InArticle Ad
- `normal` or _without value_ for Graphic and Text Ad

In case of  InArticle Ad, `width` and `height` is **not** mandatory.

> "Unique ID" have priority on "specific Adsense informations".

> The optional parameter `[class]` can be added to the tag.

### By modular integration

There are 2 ways for integration of this plugin using modular method:
- Using default configuration: User have possibility to modify and change Ad configuration through `Admin panel`.
- Using fixed configuration: User shall modify the modular page template to update Ad informations.

#### Default configuration

Modular integration using default configuration allow user to add default Ad in theming of **Grav** website reachable from Admin Panel. User have possibility to choose among 3 kinds of Ad:

- The Horizontal Ad

This way display an Ad in horizontal format by adding the following lines in your theme skeleton:

```twig
{% include 'partials/adsense-horizontal.html.twig' %}
```

Render demo:  
![Grav adsense Plugin in Horizontal](assets/img/sandy_horizontal.png)

- The Vertical Ad

This way display an Ad in vertical format by adding the following lines in your theme skeleton:

```twig
{% include 'partials/adsense-vertical.html.twig' %}
```

Render demo:  
![Grav adsense Plugin in Vertical](assets/img/sandy_vertical.png)

- The Square Ad

This way display an Ad in square format by adding the following lines in your theme skeleton:

```twig
{% include 'partials/adsense-square.html.twig' %}
```

Render demo:  
![Grav adsense Plugin in Square](assets/img/sandy_square.png)

#### Fixed configuration

Modular integration using fixed configuration allow user to add a bloc Ad in template of modular page. The following code is used:

```twig
{% set adsense_client   = "ca-pub-0000000000000000" %}
{% set adsense_slot     = "0000000000" %}
{% set adsense_width    = "721" %}
{% set adsense_height   = "90" %}

{% include 'partials/adsense.html.twig' %}
```

## Migration from 1.x to 2.x

Migration is simple: rename the file `user/config/plugins/adsense2.yaml` to `user/config/plugins/adsense.yaml`, or redo your plugin configuration.

## Contributing

You can contribute at any time but before opening any issue, please search for existing issues.

After that please note:

-   If you find a bug, would like to make a feature request or suggest an improvement, [please open a new issue][issues]. If you have any interesting ideas for additions to the syntax please do suggest them as well!
-   Feature requests are more likely to get attention if you include a clearly described use case.
-   Add images/links/etc so I can help you faster

Thanks!

## Credits

- [Adsense Grav Plugin](https://github.com/muuvmuuv/grav-plugin-adsense) of [Marvin Heilemann](https://github.com/muuvmuuv/): Initialized by Marvin H.
- [Shortcode Core](https://github.com/getgrav/grav-plugin-shortcode-core)

## License

Copyright (c) 2019 [clemdesign][github].

For use under the terms of the [MIT][mit-license] license.

[github]: https://github.com/clemdesign/ "GitHub account from Clemdesign"

[mit-license]: http://www.opensource.org/licenses/mit-license.php "MIT license"

[project]: https://github.com/clemdesign/grav-plugin-adsense

[issues]: https://github.com/clemdesign/grav-plugin-adsense/issues "GitHub Issues for Grav Adsense Plugin"


CHANGELOG.md

# v2.0.0

## 10/15/2019

1.  [](#improved)
-   Rename plugin repository `adsense2` to `adsense`.


# v1.1.4

## 10/14/2019

1.  [](#bugfix)
-   Fix issue [#6](https://github.com/clemdesign/grav-plugin-adsense/issues/6)

# v1.1.2

# v1.1.3

## 12/18/2018

1.  [](#improved)
-   Merge of Pull request [#3](https://github.com/clemdesign/grav-plugin-adsense/pull/3)

# v1.1.2

## 06/11/2018

1.  [](#bugfix)
-   Fix [Issue #2](https://github.com/clemdesign/grav-plugin-adsense/issues/2)
-   Fix issue on sandbox picture path.

# v1.1.1

## 04/05/2018

1.  [](#improved)
-   Add filter feature for Auto Ads: User have possibility to Apply Auto Ads to filtered URN items.

# v1.1.0

## 03/24/2018

1.  [](#improved)
-   Integrate Auto Ads (new feature of Adsense)

# v1.0.1

## 03/11/2018

1.  [](#bugfix)
-   Fix issue for sandbox renderer

1.  [](#improved)
-   Update README to add GPM installation
-   Update README to add Modular page - Fixed configuration method

# v1.0.0

## 02/18/2018

1.  [](#new)

-   Initiate plugin inspired by [https://github.com/muuvmuuv/grav-plugin-adsense](https://github.com/muuvmuuv/grav-plugin-adsense)


adsense.php
<?php
/**
 * This plugin enables to use AdSense inside a document
 * to be rendered by Grav.
 *
 * Licensed under the MIT Version 4 licenses, see LICENSE.
 * http://cookie-soft.de/license
 *
 * @since       1.1.0
 *
 * @see         https://github.com/muuvmuuv/grav-plugin-adsense
 *
 * @author      Marvin Heilemann <marvin.heilemann@cookie-soft.de>
 * @copyright   2015, Marvin Heilemann
 * @license     http://opensource.org/licenses/MIT
 */

namespace Grav\Plugin;

use Grav\Common\Plugin;
use Grav\Common\Page\Page;
use Grav\Common\Twig\Twig;
use RocketTheme\Toolbox\Event\Event;

/**
 * AdSensePlugin.
 *
 * This plugin enables to use AdSense inside a document
 * to be rendered by Grav.
 */
class AdSensePlugin extends Plugin
{
  /**
   * Return a list of subscribed events.
   *
   * @return array a list of events
   */
  public static function getSubscribedEvents()
  {
    return [
      'onPluginsInitialized' => ['onPluginsInitialized', 1],
    ];
  }

  /**
   * Initialize configuration.
   */
  public function onPluginsInitialized()
  {
    if ($this->isAdmin()) {
      $this->enable([
        'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
      ]);
      return;
    }

    if ($this->config->get('plugins.adsense.enabled')) {
      $this->enable([
        'onShortcodeHandlers' => ['onShortcodeHandlers', 0],
        'onPageContentRaw' => ['onPageContentRaw', 0],
        'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
        'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0],
      ]);
    }
  }

  /**
   * Initialize shortcode.
   */
  public function onShortcodeHandlers()
  {
    $this->grav['shortcode']->registerAllShortcodes(__DIR__.'/shortcodes');
  }

  /**
   * Add content after page content was read into the system.
   *
   * @param Event $event an event object, when `onPageContentRaw` is fired
   */
  public function onPageContentRaw(Event $event)
  {
    /** @var Page $page */
    $page = $event['page'];

    /** @var Twig $twig */
    $twig = $this->grav['twig'];

    $ads_config = $this->config->get("plugins.adsense.adsense");

    /* Set twig vars */
    $twig->twig_vars['adsense_sandy']     = $this->config->get('plugins.adsense.sandbox');
    $twig->twig_vars['adsense_position']  = isset($ads_config["options"]["position"]) ? $ads_config["options"]["position"]: "center";

    // Page ads
    $ad = array();
    if(isset($ads_config["page_ads"]) && is_array($ads_config["page_ads"])) {
      foreach($ads_config["page_ads"] as $value){
        if(isset($value["id"]) && strlen($value["id"])>3
          && isset($value["client"])  && strlen($value["client"])>3
          && isset($value["slot"]) && $value["slot"]>3 ) {

          $ad[$value["id"]]["client"] = $value["client"];
          $ad[$value["id"]]["slot"]   = $value["slot"];
          $ad[$value["id"]]["type"]   = $value["type"];
          $ad[$value["id"]]["height"] = isset($value["height"]) ? intval($value["height"]): 90;
          $ad[$value["id"]]["width"]  = isset($value["width"]) ? intval($value["width"]): 728;
        }
      }
    }
    $twig->twig_vars['adsense_page_ads'] = $ad;

    // Default Modular Ads
    if(isset($ads_config["modular_ads"])) {
      // Horizontal
      if(isset($ads_config["modular_ads"]["horizontal"])) {
        $modular_ad = $ads_config["modular_ads"]["horizontal"];
        if(isset($modular_ad["client"]) && strlen($modular_ad["client"])>3 &&
          isset($modular_ad["slot"]) && intval($modular_ad["slot"])>0 &&
          isset($modular_ad["width"]) && intval($modular_ad["width"])>0 &&
          isset($modular_ad["height"]) && intval($modular_ad["height"])>0) {
          $twig->twig_vars['adsense_default_horizontal_client'] = $modular_ad["client"];
          $twig->twig_vars['adsense_default_horizontal_slot'] = $modular_ad["slot"];
          $twig->twig_vars['adsense_default_horizontal_width'] = $modular_ad["width"];
          $twig->twig_vars['adsense_default_horizontal_height'] = $modular_ad["height"];
        } else {
          $twig->twig_vars['adsense_default_horizontal_client'] = null;
        }
      }
      // Vertical
      if(isset($ads_config["modular_ads"]["vertical"])) {
        $modular_ad = $ads_config["modular_ads"]["vertical"];
        if(isset($modular_ad["client"]) && strlen($modular_ad["client"])>3 &&
          isset($modular_ad["slot"]) && intval($modular_ad["slot"])>0 &&
          isset($modular_ad["width"]) && intval($modular_ad["width"])>0 &&
          isset($modular_ad["height"]) && intval($modular_ad["height"])>0) {
          $twig->twig_vars['adsense_default_vertical_client'] = $modular_ad["client"];
          $twig->twig_vars['adsense_default_vertical_slot'] = $modular_ad["slot"];
          $twig->twig_vars['adsense_default_vertical_width'] = $modular_ad["width"];
          $twig->twig_vars['adsense_default_vertical_height'] = $modular_ad["height"];
        } else {
          $twig->twig_vars['adsense_default_vertical_client'] = null;
        }
      }
      // Square
      if(isset($ads_config["modular_ads"]["square"])) {
        $modular_ad = $ads_config["modular_ads"]["square"];
        if(isset($modular_ad["client"]) && strlen($modular_ad["client"])>3 &&
          isset($modular_ad["slot"]) && intval($modular_ad["slot"])>0 &&
          isset($modular_ad["width"]) && intval($modular_ad["width"])>0 &&
          isset($modular_ad["height"]) && intval($modular_ad["height"])>0) {
          $twig->twig_vars['adsense_default_square_client'] = $modular_ad["client"];
          $twig->twig_vars['adsense_default_square_slot'] = $modular_ad["slot"];
          $twig->twig_vars['adsense_default_square_width'] = $modular_ad["width"];
          $twig->twig_vars['adsense_default_square_height'] = $modular_ad["height"];
        } else {
          $twig->twig_vars['adsense_default_square_client'] = null;
        }
      }
    }


    //Mode
    if($this->config->get('plugins.adsense.sandbox')) {
      $twig->twig_vars['adsense_mode'] = "sandy";
    } else {
      $twig->twig_vars['adsense_mode'] = isset($ads_config["options"]["mode"]) ? $ads_config["options"]["mode"]: "async";
    }


  }

  /**
   * Add style and script to page.
   */
  public function onTwigSiteVariables()
  {
    // Built In CSS
    if (!$this->isAdmin() && $this->config->get('plugins.adsense.built_in_css')) {
      /* adding the style to the assets */
      $this->grav['assets']->addCss('plugin://adsense/assets/css/adsense.css');
    }

    //Auto Ads
    if (!$this->isAdmin() && $this->config->get('plugins.adsense.use_auto_ads')) {
      $auto_ads_client = $this->config->get('plugins.adsense.adsense.options.auto_ads_client');

      if($auto_ads_client){

        //Manage Filter
        $filter_items = $this->config->get('plugins.adsense.adsense.options.auto_ads_filter');
        $url = strtolower($this->cleanUrl($this->grav["uri"]->url()));

        $bDispAutoAds = true;
        if(is_array($filter_items) && (count($filter_items) > 0)) {
          if(!empty($url)) {
            //Look for url in filter list
            if(!in_array($url, $filter_items)) $bDispAutoAds = false;
          } else {
            //Look for home in filter list
            if(!in_array("home", $filter_items)) $bDispAutoAds = false;
          }
        }

        if($bDispAutoAds) {
          //Load Google Ads in async not in pipeline
          $this->grav['assets']->addJs('//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js', 10, false, 'async');
  
          $this->grav['assets']->addInlineJs('
       (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: "'.$auto_ads_client.'",
            enable_page_level_ads: true
       });', 10, null);
        }
      }
    }

    // Editor button
    if ($this->isAdmin() && $this->config->get('plugins.adsense.add_editor_button')) {
      $this->grav['assets']->add('plugin://adsense/admin/editor-button/js/button.js');
    }
  }

  /**
   * Add current directory to twig lookup paths.
   */
  public function onTwigTemplatePaths()
  {
      $this->grav['twig']->twig_paths[] = __DIR__.'/templates';
  }

  /**
   * Clean URL
   * @param $url
   * @return bool|string
   */
  private function cleanUrl($url) {
    if(substr($url,0,1) == "/") {
      return substr($url, 1, strlen($url)-1);
    }
    return $url;
  }
}


shortcodes/AdSenseShortcode.php
<?php

namespace Grav\Plugin\Shortcodes;

use Thunder\Shortcode\Shortcode\ShortcodeInterface;

class AdSenseShortcode extends Shortcode
{
  public function init()
  {
    $this->shortcode->getHandlers()->add('adsense', function (ShortcodeInterface $sc) {

      $hash   = $this->shortcode->getId($sc);
      $adid   = $sc->getParameter("id");

      $class  = $sc->getParameter('class');

      /* Ad defined by user by integration code */
      $client = $sc->getParameter('client');
      $slot   = $sc->getParameter('slot');
      $type   = $sc->getParameter('type');
      $width  = intval($sc->getParameter('width'));
      $height = intval($sc->getParameter('height'));

      /** Load an Ad according the unique ID */
      if(!empty($adid) && isset($this->twig->twig_vars['adsense_page_ads'][$adid])){
        $params = $this->twig->twig_vars['adsense_page_ads'][$adid];

        $output = $this->twig->processTemplate('partials/adsense.html.twig', array(
          'adsense_hash'    => $hash,
          'adsense_mode'    => $this->twig->twig_vars['adsense_mode'],
          'adsense_position'=> $this->twig->twig_vars['adsense_position'],
          'adsense_client'  => $params["client"],
          'adsense_slot'    => $params["slot"],
          'adsense_height'  => $params["height"],
          'adsense_width'   => $params["width"],
          'adsense_type'    => $params["type"],
          'adsense_format'  => $this->getFormat($params["width"], $params["height"]),
          'adsense_class'   => $class,
        ));

        return $output;

        /** Load an Ad defined by user through integration shortcode */
      } elseif(isset($client) && strlen($client)>3 &&
                isset($slot) && intval($slot)>0 &&
                (($height>0 && $width>0) || (isset($type) && ($type == "inarticle" || $type == "normal")))){

        $output = $this->twig->processTemplate('partials/adsense.html.twig', array(
          'adsense_hash'    => $hash,
          'adsense_mode'    => $this->twig->twig_vars['adsense_mode'],
          'adsense_position'=> $this->twig->twig_vars['adsense_position'],
          'adsense_client'  => $client,
          'adsense_slot'    => $slot,
          'adsense_height'  => $height,
          'adsense_width'   => $width,
          'adsense_type'    => $type,
          'adsense_format'  => $this->getFormat($width , $height),
          'adsense_class'   => $class,
        ));

        return $output;
      }

      return null;
    });
  }

  /**
   * Return format for sandy box
   * @param $width
   * @param $height
   * @return string
   */
  private function getFormat($width, $height){
    if($width > $height) {
      return "horizontal";
    } else if(abs($width - $height) < 100){
      return "square";
    } else {
      return "vertical";
    }
  }
}


admin/editor-button/js/button.js
(function($){
    $(function(){
        $('body').on('grav-editor-ready', function() {
            var Instance = Grav.default.Forms.Fields.EditorField.Instance;
            Instance.addButton({
                adsense: {
                    identifier: 'adsense',
                    title: 'Adsense',
                    label: '<i class="fa fa-fw fa-adn"></i>',
                    modes: ['gfm', 'markdown'],
                    action: function(_ref) {
                        var codemirror = _ref.codemirror, button = _ref.button;
                        button.on('click.editor.adsense',function() {
                          var text = '[adsense id="" client="" slot="" width="" height="" type="inarticle/normal" class=""][/adsense]';

                          //Add text to the editor
                          var pos     = codemirror.getDoc().getCursor(true);
                          var posend  = codemirror.getDoc().getCursor(false);

                          for (var i=pos.line; i<(posend.line+1);i++) {
                            codemirror.replaceRange(text+codemirror.getLine(i), { line: i, ch: 0 }, { line: i, ch: codemirror.getLine(i).length });
                          }

                          codemirror.setCursor({ line: posend.line, ch: codemirror.getLine(posend.line).length });
                          codemirror.focus();
                        });
                    }
                }
            });
        });
    });
})(jQuery);


adsense2.php
<?php
/**
 * This plugin enables to use AdSense inside a document
 * to be rendered by Grav.
 *
 * Licensed under the MIT Version 4 licenses, see LICENSE.
 * http://cookie-soft.de/license
 *
 * @since       1.1.0
 *
 * @see         https://github.com/muuvmuuv/grav-plugin-adsense
 *
 * @author      Marvin Heilemann <marvin.heilemann@cookie-soft.de>
 * @copyright   2015, Marvin Heilemann
 * @license     http://opensource.org/licenses/MIT
 */

namespace Grav\Plugin;

use Grav\Common\Plugin;
use Grav\Common\Page\Page;
use Grav\Common\Twig\Twig;
use RocketTheme\Toolbox\Event\Event;

/**
 * AdSensePlugin.
 *
 * This plugin enables to use AdSense inside a document
 * to be rendered by Grav.
 */
class AdSense2Plugin extends Plugin
{
  /**
   * Return a list of subscribed events.
   *
   * @return array a list of events
   */
  public static function getSubscribedEvents()
  {
    return [
      'onPluginsInitialized' => ['onPluginsInitialized', 1],
    ];
  }

  /**
   * Initialize configuration.
   */
  public function onPluginsInitialized()
  {
    if ($this->isAdmin()) {
      $this->enable([
        'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
      ]);
      return;
    }

    if ($this->config->get('plugins.adsense2.enabled')) {
      $this->enable([
        'onShortcodeHandlers' => ['onShortcodeHandlers', 0],
        'onPageContentRaw' => ['onPageContentRaw', 0],
        'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
        'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0],
      ]);
    }
  }

  /**
   * Initialize shortcode.
   */
  public function onShortcodeHandlers()
  {
    $this->grav['shortcode']->registerAllShortcodes(__DIR__.'/shortcodes');
  }

  /**
   * Add content after page content was read into the system.
   *
   * @param Event $event an event object, when `onPageContentRaw` is fired
   */
  public function onPageContentRaw(Event $event)
  {
    /** @var Page $page */
    $page = $event['page'];

    /** @var Twig $twig */
    $twig = $this->grav['twig'];

    $ads_config = $this->config->get("plugins.adsense2.adsense");

    /* Set twig vars */
    $twig->twig_vars['adsense_sandy']     = $this->config->get('plugins.adsense2.sandbox');
    $twig->twig_vars['adsense_position']  = isset($ads_config["options"]["position"]) ? $ads_config["options"]["position"]: "center";

    // Page ads
    $ad = array();
    if(isset($ads_config["page_ads"]) && is_array($ads_config["page_ads"])) {
      foreach($ads_config["page_ads"] as $value){
        if(isset($value["id"]) && strlen($value["id"])>3
          && isset($value["client"])  && strlen($value["client"])>3
          && isset($value["slot"]) && $value["slot"]>3 ) {

          $ad[$value["id"]]["client"] = $value["client"];
          $ad[$value["id"]]["slot"]   = $value["slot"];
          $ad[$value["id"]]["type"]   = $value["type"];
          $ad[$value["id"]]["height"] = isset($value["height"]) ? intval($value["height"]): 90;
          $ad[$value["id"]]["width"]  = isset($value["width"]) ? intval($value["width"]): 728;
        }
      }
    }
    $twig->twig_vars['adsense_page_ads'] = $ad;

    // Default Modular Ads
    if(isset($ads_config["modular_ads"])) {
      // Horizontal
      if(isset($ads_config["modular_ads"]["horizontal"])) {
        $modular_ad = $ads_config["modular_ads"]["horizontal"];
        if(isset($modular_ad["client"]) && strlen($modular_ad["client"])>3 &&
          isset($modular_ad["slot"]) && intval($modular_ad["slot"])>0 &&
          isset($modular_ad["width"]) && intval($modular_ad["width"])>0 &&
          isset($modular_ad["height"]) && intval($modular_ad["height"])>0) {
          $twig->twig_vars['adsense_default_horizontal_client'] = $modular_ad["client"];
          $twig->twig_vars['adsense_default_horizontal_slot'] = $modular_ad["slot"];
          $twig->twig_vars['adsense_default_horizontal_width'] = $modular_ad["width"];
          $twig->twig_vars['adsense_default_horizontal_height'] = $modular_ad["height"];
        } else {
          $twig->twig_vars['adsense_default_horizontal_client'] = null;
        }
      }
      // Vertical
      if(isset($ads_config["modular_ads"]["vertical"])) {
        $modular_ad = $ads_config["modular_ads"]["vertical"];
        if(isset($modular_ad["client"]) && strlen($modular_ad["client"])>3 &&
          isset($modular_ad["slot"]) && intval($modular_ad["slot"])>0 &&
          isset($modular_ad["width"]) && intval($modular_ad["width"])>0 &&
          isset($modular_ad["height"]) && intval($modular_ad["height"])>0) {
          $twig->twig_vars['adsense_default_vertical_client'] = $modular_ad["client"];
          $twig->twig_vars['adsense_default_vertical_slot'] = $modular_ad["slot"];
          $twig->twig_vars['adsense_default_vertical_width'] = $modular_ad["width"];
          $twig->twig_vars['adsense_default_vertical_height'] = $modular_ad["height"];
        } else {
          $twig->twig_vars['adsense_default_vertical_client'] = null;
        }
      }
      // Square
      if(isset($ads_config["modular_ads"]["square"])) {
        $modular_ad = $ads_config["modular_ads"]["square"];
        if(isset($modular_ad["client"]) && strlen($modular_ad["client"])>3 &&
          isset($modular_ad["slot"]) && intval($modular_ad["slot"])>0 &&
          isset($modular_ad["width"]) && intval($modular_ad["width"])>0 &&
          isset($modular_ad["height"]) && intval($modular_ad["height"])>0) {
          $twig->twig_vars['adsense_default_square_client'] = $modular_ad["client"];
          $twig->twig_vars['adsense_default_square_slot'] = $modular_ad["slot"];
          $twig->twig_vars['adsense_default_square_width'] = $modular_ad["width"];
          $twig->twig_vars['adsense_default_square_height'] = $modular_ad["height"];
        } else {
          $twig->twig_vars['adsense_default_square_client'] = null;
        }
      }
    }


    //Mode
    if($this->config->get('plugins.adsense2.sandbox')) {
      $twig->twig_vars['adsense_mode'] = "sandy";
    } else {
      $twig->twig_vars['adsense_mode'] = isset($ads_config["options"]["mode"]) ? $ads_config["options"]["mode"]: "async";
    }


  }

  /**
   * Add style and script to page.
   */
  public function onTwigSiteVariables()
  {
    // Built In CSS
    if (!$this->isAdmin() && $this->config->get('plugins.adsense2.built_in_css')) {
      /* adding the style to the assets */
      $this->grav['assets']->addCss('plugin://adsense2/assets/css/adsense.css');
    }

    //Auto Ads
    if (!$this->isAdmin() && $this->config->get('plugins.adsense2.use_auto_ads')) {
      $auto_ads_client = $this->config->get('plugins.adsense2.adsense.options.auto_ads_client');

      if($auto_ads_client){

        //Manage Filter
        $filter_items = $this->config->get('plugins.adsense2.adsense.options.auto_ads_filter');
        $url = strtolower($this->cleanUrl($this->grav["uri"]->url()));

        $bDispAutoAds = true;
        if(is_array($filter_items) && (count($filter_items) > 0)) {
          if(!empty($url)) {
            //Look for url in filter list
            if(!in_array($url, $filter_items)) $bDispAutoAds = false;
          } else {
            //Look for home in filter list
            if(!in_array("home", $filter_items)) $bDispAutoAds = false;
          }
        }

        if($bDispAutoAds) {
          //Load Google Ads in async not in pipeline
          $this->grav['assets']->addJs('//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js', 10, false, 'async');
  
          $this->grav['assets']->addInlineJs('
       (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: "'.$auto_ads_client.'",
            enable_page_level_ads: true
       });', 10, null);
        }
      }
    }

    // Editor button
    if ($this->isAdmin() && $this->config->get('plugins.adsense2.add_editor_button')) {
      $this->grav['assets']->add('plugin://adsense2/admin/editor-button/js/button.js');
    }
  }

  /**
   * Add current directory to twig lookup paths.
   */
  public function onTwigTemplatePaths()
  {
      $this->grav['twig']->twig_paths[] = __DIR__.'/templates';
  }

  /**
   * Clean URL
   * @param $url
   * @return bool|string
   */
  private function cleanUrl($url) {
    if(substr($url,0,1) == "/") {
      return substr($url, 1, strlen($url)-1);
    }
    return $url;
  }
}
